pyquil/kraus.py

```diff
@@ -56,3 +58,211 @@ def _create_kraus_pragmas(name, qubit_indices, kraus_ops):
+    :return: A list of transformed Kraus operators.
+    """
+    return [kj.dot(gate_matrix) for kj in kraus_ops]
+
+
+def damping_kraus_map(p=0.10):
+    """
+    Generate the Kraus operators corresponding to an amplitude damping
+    noise channel.
+
+    :param float p: The one-step damping probability.
+    :return: A list [k1, k2] of the Kraus operators that parametrize the map.
+    :rtype: list
+    """
+    damping_op = np.sqrt(p) * np.array([[0, 1],
+                                        [0, 0]])
+
+    residual_kraus = np.diag([1, np.sqrt(1 - p)])
+    return [residual_kraus, damping_op]
+
+
+def dephasing_kraus_map(p=0.10):
+    """
+    Generate the Kraus operators corresponding to a dephasing channel.
+
+    :params float p: The one-step dephasing probability.
+    :return: A list [k1, k2] of the Kraus operators that parametrize the map.
+    :rtype: list
+    """
+    return [np.sqrt(1 - p) * np.eye(2), np.sqrt(p) * np.diag([1, -1])]
+
+
+def tensor_kraus_maps(k1, k2):
+    """
+    Generate the Kraus map corresponding to the composition
+    of two maps on different qubits.
+
+    :param list k1: The Kraus operators for the first qubit.
+    :param list k2: The Kraus operators for the second qubit.
+    :return: A list of tensored Kraus operators.
+    """
+    return [np.kron(k1j, k2l) for k1j in k1 for k2l in k2]
+
+
+def combine_kraus_maps(k1, k2):
+    """
+    Generate the Kraus map corresponding to the composition
+    of two maps on the same qubits with k1 being applied to the state
+    after k2.
+
+    :param list k1: The list of Kraus operators that are applied second.
+    :param list k2: The list of Kraus operators that are applied first.
+    :return: A combinatorically generated list of composed Kraus operators.
+    """
+    return [np.dot(k1j, k2l) for k1j in k1 for k2l in k2]
+
+
+def damping_after_dephasing(T1, T2, gate_time):
+    """
+    Generate the Kraus map corresponding to the composition
+    of a dephasing channel followed by an amplitude damping channel.
+
+    :param float T1: The amplitude damping time
+    :param float T2: The dephasing time
+    :param float gate_time: The gate duration.
+    :return: A list of Kraus operators.
+    """
+    damping = damping_kraus_map(p=gate_time / T1)
+    dephasing = dephasing_kraus_map(p=gate_time / T2)
+    return combine_kraus_maps(damping, dephasing)
+
+
+# You can only apply gate-noise to non-parametrized gates,
+# so we need to define placeholders for RX(+/- pi/2)
+SINGLE_Q = {
+    "noisy-x-plus90": expm(-1j * np.pi / 4 * np.array([[0, 1],
+                                                       [1, 0]])),
+    "noisy-x-minus90": expm(+1j * np.pi / 4 * np.array([[0, 1],
+                                                        [1, 0]])),
+}
+
+TWO_Q = {
+    "noisy-cz": np.diag([1, 1, 1, -1]),
+}
+
+
+def noisy_instruction(instruction):
+    if isinstance(instruction, Pragma):
+        return instruction
+
+    if instruction.name == 'RZ':
```

Comment:
You should be checking if the instruction is a Gate and not something else

Reply:
Are there other types of instructions? How defensive should I be? I'm thinking adding: 

```python
if not istinstance(instruction, Gate):
    warnings.warn(f"Unknown instruction type for {instruction}. Noise might not make sense.")
    return instruction
```

(after the check that if it's a pragma)

Reply:
There are a bunch of other instructions, Measure is the most common but there's also control sequence instructions like labels and jumps and whatever.

Why not throw out programs that can't be made noisy? I think there's a method "is_protoquil" or something on Programs that could be useful here.

Reply:
I think we generally do not want to restrict the type of programs that can be simulated on a noisy qvm, e.g., we might want to assume some noise model and then simulate an error correction scheme that is expressed as Quil program.

Reply:
In the spirit of this high-level API, I'd be more inclined to do as @stevenheidel suggested and be pretty picky with what this function takes. If you need something more complicated, you can use the lower level functions. Also, we can always relax restrictions as they become a problem whereas the other way around might involve the noise model silently not doing what the user expects.

Reply:
Or give a warning for anything that's not a gate

Reply:
Hmm, so I understand if there is some concern about complicated constructs such as `DEFCIRCUIT`, but as far as I can tell there is currently no substantial support for these anyway. The noise macros should not interact in any unexpected way with JUMP(-WHEN|UNLESS) statements as well as LABEL's and MEASURE's, so let's at least pass those through this function without alteration (or warning).

Reply:
Ok, that sounds good. Actually it got me thinking: should we warn users that this code *doesn't* simulate measurement error? Maybe I'll put something in the docstrings

Reply:
I think it's not an absolute necessity. Noise models for readout will be added very soon and Kraus maps are explicit models for gate operations not readout so I am semi-hopeful that nobody will be confused about this in the interim.